### DISCUSSION TITLE
execute_if_off ??

### DISCUSSION THREAD

**Discussion by agsola** [Koenkk/zigbee2mqtt#13015] [General]
I'm trying to understand who to use execute_if_off.
I can't find in any device, I can't find in the configuration.

However, I've seen in different issues like:
https://github.com/Koenkk/zigbee-herdsman-converters/issues/2284
https://github.com/Koenkk/zigbee-herdsman-converters/issues/4192
https://github.com/nurikk/zigbee2mqtt-frontend/issues/356

I have some LED drivers and IKEA Tradfri bulbs.
In the LED when I change the light temperature while the device is turned off, when turning on the temperature is changed.
However, in the Tradfri bulbs, when I change the temperature while the bulbs are off, when turning on the don't change, and they keep the last temperature value.
However, none of them expose the "execute_if_off".

**Comment by sjorge** [comment]
You are correct that some level and color options are not in the expose data (anymore).

If a bulb supports it or not is a mixed bag, Trådfri bulbs on the ZB3 firmware and Hue (With BT) seem to support it though, there is also still a weird rendering issue with expose things with subkeys in the frontend is still a problem too.

Currently for color_options only the `execute_if_off` bit has support, although this could be expanded to do some of the color_temp linking to brightness options in the future, I never got them to work so I didn't have anything to test with so they are currently not implemented.

There is no separate level_options, the `execute_if_off` for that got implemented under level_config which sets a lot of other things too. 

I'm not sure which is the better approach tbh, `level_config` has a bunch of stuff as subkeys, include `execute_if_off` which get written to the options attribute, it would get complicated if other options bit-fields ever get implemented, less so for `color_options` as this is a separate property. But that does mean things like the color_temp_startup are separate properties also instead of all living under a `color_config` property as sub keys.

History aside, you can configure it like this:
```
mosquitto_pub -t zigbee2mqtt/<friendly_name>/set/color_options -m '{"execute_if_off": true}'
mosquitto_pub -t zigbee2mqtt/<friendly_name>/set/level_config -m '{"execute_if_off": true}'
```

When enabled (and if the bulb supports it) you can now change the color_temp, color, brightness without the bulb being on or turning on when doing so.

**Comment by agsola** [comment]
Thank you so much @sjorge !!
execute_if_off works great for color temperature, however, with brightness options I can't get it to work.

I had two different brands, Tradfri bulbs and Iluminize LED drivers. Iluminize by default were able to execute_if_off, now all my devices are able to execute if off. 

Regarding the brigthness, I've tried first to send the commands of color_options and level_config.
After that, I've tried with some commands like:

```
mosquitto_pub -t zigbee2mqtt/<friendly_name>/set -m '{"brightness":15}'
mosquitto_pub -t zigbee2mqtt/<friendly_name>/set -m '{"level_config":{"execute_if_off":true},"brightness":15}'
mosquitto_pub -t zigbee2mqtt/<friendly_name>/set -m '{"level_config":{"execute_if_off":true},"brightness":15, "state":"OFF"}'

```

But none of them worked. All of them turn the light on except for the last one that only turned off the light, ignoring the brightness. So I find no option to change the brightness while keeping the light off. Any idea of how to do it?

**Comment by sjorge** [comment]
Interesting, I have the same result... I did some more poking around the spec

<img width="865" alt="image" src="https://user-images.githubusercontent.com/379665/177032953-1898cdf3-91bf-43c1-8ed7-0de453f9a119.png">

I think we're hitting this in the longer list of conditions: `The On/Off cluster exists on the same endpoint as this cluster.`

All the bulbs I have have an genOnOff cluster on the same endpoint as the genLevelCtrl.

cc: @Koenkk as your a bit better than me at deciphering the vagueness/contraditions of the ZCL at times.

**Comment by agsola** [comment]
It makes sense. According to the specs, this shouldn't work right now.
Could it make sense to use a separate cluster to change color brightness with no "On/Off cluster"?

**Comment by JeanneD4RK** [comment]
What I understand in the doc is that the command shloud continue since all of the critera are not met, because the last one is about ExecuteIfOff but is not 0.

This issue is still present as of today.

**Comment by thegcat** [comment]
Has anyone found a solution to change the brightness of Tradfri (or Kajplats) bulbs while off?

**Comment by Nama** [comment]
so whats the workaround?
check the group state, if on, loop thru all devices and check which is on and send the command only to that device?
doesnt seem feasible...

**Comment by andrei-lazarov** [comment]
Go to device -> dev console -> read / write attributes
- cluster `genLevelCtrl` -> attribute `options` -> 1 = allow changing brightness while off, 0 = don't allow
- cluster `lightingColorCtrl` -> attribute `options` -> 1 = allow changing color/temp while off, 0 = don't allow

These options are also exposed as `execute_if_off` in the UI for some devices (e.g. IKEA bulbs).

Then send MQTT to `friendly_name/set` with payload:
- `{"brightness": 254}` - turn on/off and change brightness (max 254)
- `{"brightness": 254, "state": null}` - don't turn on/off, only change brightness (works only when `execute_if_off` is enabled)

Note that not all devices support this / handle this well.
Also my IKEA bulbs seem to forget the option after a while 😭?? I need to set it again, every 2 weeks or so

**Reply by agsola** [reply]
On my old IKEA bulbs, on all of them, it was working for me before for colour, it is not working anymore, despite having the execute_if_off on both for colour and brightness. 

However, I have a new one that replaced a broken one, and it works just fine.

Any idea?

**Reply by andrei-lazarov** [reply]
Disable `execute_if_off` and enable it again

**Reply by agsola** [reply]
I've tried that, also sending the command to the device itself, to a group that the device belongs, etc. Nothing works. But everthing worked before.

**Comment by Nama** [comment]
oh my god, i was missing `"state": null`.
thank you very much!!

**Reply by agsola** [reply]
On my old IKEA bulbs, on all of them, it was working for me before for colour, it is not working anymore, despite having the execute_if_off on both for colour and brightness. 

However, I have a new one that replaced a broken one, and it works just fine.

Any idea?

**Reply by Nama** [reply]
no idea.
my bulbs are like five years old
